src/protocols/bdx/BdxMessages.cpp

```diff
@@ -0,0 +1,456 @@
+        {
+            aBuffer.PutLE32(static_cast<uint32_t>(StartOffset));
+        }
+    }
+
+    if (MaxLength > 0)
+    {
+        if (widerange)
+        {
+            aBuffer.PutLE64(MaxLength);
+        }
+        else
+        {
+            aBuffer.PutLE32(static_cast<uint32_t>(MaxLength));
+        }
+    }
+
+    aBuffer.PutLE16(FileDesLength);
+    if (FileDesignator)
+    {
+        aBuffer.Put(FileDesignator, static_cast<size_t>(FileDesLength));
+    }
+
+    if (Metadata)
+    {
+        aBuffer.Put(Metadata, static_cast<size_t>(MetadataLength));
+    }
+}
+
+CHIP_ERROR TransferInit::Parse(const System::PacketBuffer & aBuffer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    uint8_t proposedTransferCtl;
+    uint8_t rangeCtl;
+    uint32_t tmpUint32Value = 0; // Used for reading non-wide length and offset fields
+    uint8_t * bufStart      = aBuffer.Start();
+    Reader bufReader(bufStart, aBuffer.DataLength());
+    BitFlags<uint8_t, RangeControlFlags> rangeCtlFlags;
+
+    SuccessOrExit(bufReader.Read8(&proposedTransferCtl).Read8(&rangeCtl).Read16(&MaxBlockSize).StatusCode());
+
+    Version = proposedTransferCtl & kVersionMask;
+    TransferCtlOptions.SetRaw(static_cast<uint8_t>(proposedTransferCtl & ~kVersionMask));
+    rangeCtlFlags.SetRaw(rangeCtl);
+
+    if (rangeCtlFlags.Has(kStartOffset))
+    {
+        if (rangeCtlFlags.Has(kWiderange))
+        {
+            SuccessOrExit(bufReader.Read64(&StartOffset).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            StartOffset = tmpUint32Value;
+        }
+    }
+
+    if (rangeCtlFlags.Has(kDefLen))
+    {
+        if (rangeCtlFlags.Has(kWiderange))
+        {
+            SuccessOrExit(bufReader.Read64(&MaxLength).StatusCode());
+        }
+        else
+        {
+            SuccessOrExit(bufReader.Read32(&tmpUint32Value).StatusCode());
+            MaxLength = tmpUint32Value;
+        }
+    }
+
+    SuccessOrExit(bufReader.Read16(&FileDesLength).StatusCode());
+
+    // WARNING: this struct will store a pointer to the start of the file designator in the PacketBuffer,
+    // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+    // or there is risk of unsafe memory access.
+    VerifyOrExit(bufReader.HasAtLeast(FileDesLength), err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    FileDesignator = &bufStart[bufReader.OctetsRead()];
+
+    // Rest of message is metadata (could be empty)
+    if (bufReader.Remaining() > FileDesLength)
+    {
+        // WARNING: this struct will store a pointer to the start of metadata in the PacketBuffer,
+        // but will not make a copy. It is essential that this struct not outlive the PacketBuffer,
+        // or there is risk of unsafe memory access.
+        uint16_t metadataStartIndex = static_cast<uint16_t>(bufReader.OctetsRead() + FileDesLength);
+        Metadata                    = &bufStart[metadataStartIndex];
```

Comment:
Before the if(), or in an else, suggest adding:

```
  Metadata = nullptr;
  MetadataLength = 0;
```

This would handle the case of Parse having been called in twice separately on same object with different source payload.

Overall, every "unpack"  needs to set *every* field of the struct by the time a success status is returned.

Reply:
Good catch, thanks.